File: 3739-count-subarrays-with-majority-element-ii/3739-count-subarrays-with-majority-element-ii.py
class Solution:
    def countMajoritySubarrays(self, nums: List[int], target: int) -> int:
        n=len(nums)
        cnt_subarray=0
        fr=[0]*(2*n+1)
        curr_sum=0
        fr[0+n]=1
        small_cnt=0
        for i in range(n):
            if nums[i]==target :
                small_cnt+=fr[curr_sum+n]
                curr_sum+=1
            else :
                curr_sum-=1
                small_cnt-=fr[curr_sum+n]
            cnt_subarray+=small_cnt
            fr[curr_sum+n]+=1
        return cnt_subarray

File: 3739-count-subarrays-with-majority-element-ii/3739-count-subarrays-with-majority-element-ii.cpp
class Solution {
public:
    long long countMajoritySubarrays(vector<int>& nums, int target) {
        long long cnt_subarray=0;
        int n=nums.size();
        vector<int> fr(2*n+1,0);
        int curr_sum=0;
        fr[0+n]=1;
        long long small_cnt=0;
        for(int i=0;i<n;i++){
            if(nums[i]==target){
                small_cnt+=fr[curr_sum+n];
                curr_sum+=1;
            }
            else{
                curr_sum-=1;
                small_cnt-=fr[curr_sum+n];
            }
            cnt_subarray+=small_cnt;
            fr[curr_sum+n]++;
        }
        return cnt_subarray;
    }
};